server/src/dependency/rocksdb.rs

```diff
@@ -209,10 +211,20 @@ impl<'db> umi_app::Dependencies<'db> for RocksDbReaderDependencies {
 
     fn on_payload() -> &'db Self::OnPayload {
         &|state, id, hash| {
-            state.payload_queries.add_block_hash(id, hash).map_err(|e| {
-                tracing::error!("on_payload callback failed: {e:?}");
-                Error::DatabaseState
-            })
+            state
+                .payload_queries
+                .add_block_hash(id, hash)
+                .map_err(|e| {
+                    tracing::error!("on_payload callback failed: {e:?}");
+                    Error::DatabaseState
+                })?;
+            state
+                .state_queries
+                .push_state_root(hash, state.state.state_root())
+                .map_err(|e| {
+                    tracing::error!("on_payload callback failed: {e:?}");
+                    Error::DatabaseState
+                })
```

Comment:
I believe this could interfere with withdrawals root setting after OP upgrade. Previously the state root was updated right after executing the transactions, and before the header was constructed or the storage root of `L2ToL1MessagePasser` queried. But the `on_payload()` callback is applied much later down the line, so the state root at the time of setting might be stale.

Another thing I'm not sure about here is whether resolver needs to be updated too, considering that withdrawals root calculation relies on the EVM DB, virtually making it equivalent to the `state::switch_state_root()` call as is done in `command::forkchoice_update()` itself.